specs/psycopg_analysis.md

```diff
@@ -0,0 +1,176 @@
+
+Currently, Pyright assumes that any package marked as "py.typed" will provide unambiguous type declarations for all of its public symbols. Any symbols that are left unannotated (with a few specific exceptions spelled out in the documentation) are interpreted by Pyright as `Unknown`. The intent behind this approach is to increase the visibility of the missing type annotations so library authors can take action and eliminate the holes.
+
+While this approach has proven successful at raising the visibility with library authors, it has two significant downsides:
+
+1. It effectively "punishes" users of Pyright and Pylance who consume "py.typed" libraries when type annotations are missing. When referencing these unannotated symbols, they receive no completion suggestions, and they receive difficult-to-correct errors when referencing these symbols in "strict" static type checking mode. Users have little or no recourse other than to report the problem to the library maintainer and hope that it is addressed in the next package update.
+2. Some library maintainers are reluctant to add explicit type annotations in cases where they feel there is low risk of type inference ambiguity. This leaves them in an awkward situation when consumers of their library express dissatisfaction.
+
+It is perhaps time to consider an alternative approach, one that better balances the interests of users and library authors.
+
+## Recommendations
+
+I'm recommending the following changes:
+
+### Pyright's treatment of unannotated symbols in "py.typed" libraries
+
+Today, Pyright treats unannotated symbols in a "py.typed" library as "Unknown" if type checking is enabled. When typeCheckingMode is "off", it falls back to its inference logic. This effectively "punishes" users who want to use Pyright and Pylance for static type checking. I recommend that we change this behavior and always fall back on type inference logic when annotations are not present. This change will unfortunately reduce the visibility of missing annotations, so I worry that it will slow efforts to improve type completeness and consistency across the Python ecosystem, but I think it represents a pragmatic tradeoff.
+
+As a mitigation for this lack of visibility, we might want to modify Pyright and Pylance to display "ambiguous types" in a way that differentiates them from "unambiguous types". For example, we could prepend a `~` character to indicate that the type originated from a "py.typed" library and was inferred in a way that might be ambiguous.
```

Comment:
I'm not convinced that most users will care about consistency so much that a prefix is useful. For example, mypy prepends a `*` prefix in front of inferred types, and nobody seems to find it useful: https://github.com/python/mypy/issues/10076

Reply:
I see one place where it can matter. Should type errors be reported for ambiguous types? A fair amount of false positives get produced if you let pyright infer types on tensorflow mostly because of it's dynamic magic.

It'd be ideal for me if ambiguous types had hover/suggestions but did not trigger errors. Although if it's possible to disable this inference with `useLibraryCodeForTypes ` I could also just continue to set that to False.

Reply:
My concern goes away given this is restricted to py.typed. I would expect a library that claims to be typed to have minimal false positives from ambiguous errors or be willing to improve them at a reasonable pace. If a library has a lot dynamic magic in public api and doesn't type hint/stub it appropriately then I think it's misclaiming to be "py.typed".